pkg/reconciler/ingress/controller.go

```diff
@@ -0,0 +1,112 @@
+	"knative.dev/networking/pkg/status"
+	endpointsinformer "knative.dev/pkg/client/injection/kube/informers/core/v1/endpoints"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+
+	gwapiclient "knative.dev/net-gateway-api/pkg/client/gatewayapi/injection/client"
+	gatewayinformer "knative.dev/net-gateway-api/pkg/client/gatewayapi/injection/informers/apis/v1alpha1/gateway"
+	httprouteinformer "knative.dev/net-gateway-api/pkg/client/gatewayapi/injection/informers/apis/v1alpha1/httproute"
+	"knative.dev/net-gateway-api/pkg/reconciler/ingress/config"
+)
+
+// NewController initializes the controller and is called by the generated code
+// Registers eventhandlers to enqueue events
+func NewController(
+	ctx context.Context,
+	cmw configmap.Watcher,
+) *controller.Impl {
+	logger := logging.FromContext(ctx)
+
+	ingressInformer := ingressinformer.Get(ctx)
+	httprouteInformer := httprouteinformer.Get(ctx)
+	gatewayInformer := gatewayinformer.Get(ctx)
+	endpointsInformer := endpointsinformer.Get(ctx)
+
+	c := &Reconciler{
+		gwapiclient:     gwapiclient.Get(ctx),
+		httprouteLister: httprouteInformer.Lister(),
+	}
+
+	filterFunc := reconciler.AnnotationFilterFunc(networking.IngressClassAnnotationKey, GatewayAPIIngressClassName, true)
+
+	impl := ingressreconciler.NewImpl(ctx, c, GatewayAPIIngressClassName, func(impl *controller.Impl) controller.Options {
+		configsToResync := []interface{}{
+			&network.Config{},
+			&config.Gateway{},
+		}
+		resync := configmap.TypeFilter(configsToResync...)(func(string, interface{}) {
+			impl.GlobalResync(ingressInformer.Informer())
+		})
+		configStore := config.NewStore(logging.WithLogger(ctx, logger.Named("config-store")), resync)
+		configStore.WatchConfigs(cmw)
+		return controller.Options{
+			ConfigStore:       configStore,
+			PromoteFilterFunc: filterFunc,
+		}
+	})
+
+	logger.Info("Setting up Ingress event handlers")
+	ingressHandler := cache.FilteringResourceEventHandler{
+		FilterFunc: filterFunc,
+		Handler:    controller.HandleAll(impl.Enqueue),
+	}
+
+	ingressInformer.Informer().AddEventHandler(ingressHandler)
+
+	httprouteInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: filterFunc,
+		Handler:    controller.HandleAll(impl.EnqueueControllerOf),
+	})
+	gatewayInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: filterFunc,
+		Handler:    controller.HandleAll(impl.EnqueueControllerOf),
+	})
+
+	statusProber := status.NewProber(
+		logger.Named("status-manager"),
+		NewProbeTargetLister(logger, endpointsInformer.Lister()),
+		func(ing *v1alpha1.Ingress) {
+			logger.Debugf("Ready callback triggered for ingress: %s/%s", ing.Namespace, ing.Name)
+			impl.EnqueueKey(types.NamespacedName{Namespace: ing.Namespace, Name: ing.Name})
+		})
+	c.statusManager = statusProber
+	statusProber.Start(ctx.Done())
```

Comment:
Are there any discussions in the gateway-api community to make probing unnecessary? IIRC they wanted to surface a consistent status that'd do that. If so, could we add a TODO here to watch that work upstream (with a link at best)?